pkg/reconciler/clusterducktype/clusterducktype.go

```diff
@@ -18,37 +18,127 @@ package clusterducktype
 
 import (
 	"context"
+	"sort"
 
-	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/client/listers/apiextensions/v1"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	apiextensionslisters "k8s.io/apiextensions-apiserver/pkg/client/listers/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/labels"
 
-	discoveryv1alpha1 "knative.dev/discovery/pkg/apis/discovery/v1alpha1"
+	v1alpha1 "knative.dev/discovery/pkg/apis/discovery/v1alpha1"
 	ducktypereconciler "knative.dev/discovery/pkg/client/injection/reconciler/discovery/v1alpha1/clusterducktype"
-	"knative.dev/pkg/logging"
 	"knative.dev/pkg/reconciler"
 )
 
 // Reconciler implements ducktypereconciler.Interface for
 // ClusterDuckType resources.
 type Reconciler struct {
-	crdLister apiextensionsv1.CustomResourceDefinitionLister
+	crdLister apiextensionslisters.CustomResourceDefinitionLister
 }
 
 // Check that our Reconciler implements Interface
 var _ ducktypereconciler.Interface = (*Reconciler)(nil)
 
-// ReconcileKind implements Interface.ReconcileKind.
-func (r *Reconciler) ReconcileKind(ctx context.Context, o *discoveryv1alpha1.ClusterDuckType) reconciler.Event {
-	logger := logging.FromContext(ctx)
+// ReconcileKind implements Interface
+func (r *Reconciler) ReconcileKind(ctx context.Context, dt *v1alpha1.ClusterDuckType) reconciler.Event {
+	/// By query
 
-	crd, err := r.crdLister.Get("channels.messaging.knative.dev") // TODO generalize past testing
-	if err != nil {
-		logger.Errorf("error getting crd: %q", err)
-	} else {
-		logger.Errorf("found crd: %q", crd)
+	kinds := make(map[string]*apiextensionsv1.CustomResourceDefinition, 0)
+
+	for _, st := range dt.Spec.Selectors {
+		crds, err := r.getCRDsWith(st.LabelSelector)
+		if err != nil {
+			// TODO: this should be a condition that reports back that RBAC is incorrect for getting CRDs?
+			return err
+		}
+		for _, crd := range crds {
+			key := crd.Name
+			if _, found := kinds[key]; !found {
+				kinds[key] = crd
+			}
+		}
+	}
+
+	foundDucks := make([]v1alpha1.FoundDuck, 0)
+	for _, crd := range kinds {
+		foundDucks = append(foundDucks, CRDToFoundDuck("", crd))
+	}
+
+	/// By ref
+
+	for _, dv := range dt.Spec.Versions {
+		for _, ref := range dv.Refs {
+			// TODO we should query and test that the Ref is installed and works on this cluster.
+			foundDucks = append(foundDucks, v1alpha1.FoundDuck{
```

Comment:
Sorry, I'm missing some context, we have a type called FoundDuck? What is it for? The name sounded weird, especially the Found part...

Reply:
We can bikeshed on the name, but it is a touple of `duck version` and `ref`

Reply:
ack. I guess I would have preferred just Duck.
Yeah, you can change it later on if you wish...

Reply:
if you have a good idea, let me know! It is only used in the status list of ducks.

Reply:
I am not sold yet on the shape of the status, it is adapted from the original prototype, and looks like this:
```
status:
  ducks:
  - duckVersion: v1
    ref:
      group: apps
      kind: ReplicaSet
      version: v1
  - duckVersion: v1
    ref:
      group: apps
      kind: DaemonSet
      version: v1
```

But what do you think of this:

```
status:
  ducks:
    v1:
    - group: apps
       kind: ReplicaSet
       version: v1
    - group: apps
       kind: DaemonSet
       version: v1
```

so the change would be status.ducks == map[string]GVK where string is version of duck

Reply:
I think the second option seems cleaner.
Does ducks v1 can have GVKs that are not v1s? If not, then you might just have GKs?

Do you care about the grouping based on versions? If not, then maybe a list of GVKs would do?
```
status:
  ducks:
    - group: apps
       kind: DaemonSet
       version: v1
    - group: apps
       kind: DaemonSet
       version: v1beta1
```

We can get this in, and then you make another pass changing the shape if you think it can be improved.

Reply:
We need to know the duck version and how that relates to the CRD version, that is the point of all this madness...